tool/embedded-binaries.ts

```diff
@@ -70,162 +71,216 @@ export async function getEmbeddedProtocol(outPath: string): Promise<void> {
+    throw Error(
+      `Failed to unzip ${options.repo} release asset: ${error.message}.`
+    );
+  }
+
+  return releaseInfo.name.replace(' ', '-');
+}
+
+// Clones `repo` into `outPath`, then checks out the given Git `ref`.
+function fetchRepo(repo: string, outPath: string, ref?: string) {
+  if (!existsSync(p.join(outPath, repo))) {
+    console.log(`Cloning ${repo} into ${outPath}.`);
+    shell.exec(
+      `git clone \
+      --depth=1 \
+      git://github.com/sass/${repo} \
+      ${p.join(outPath, repo)}`,
+      {silent: true}
+    );
+  }
+
+  const version = ref ? `commit ${ref}` : 'latest update';
+  console.log(`Fetching ${version} for ${repo}.`);
+  shell.exec(`git fetch --depth=1 origin ${ref ?? 'master'}`, {
+    silent: true,
+    cwd: p.join(outPath, repo),
+  });
+  shell.exec('git reset --hard FETCH_HEAD', {
+    silent: true,
+    cwd: p.join(outPath, repo),
+  });
+}
+
+// Links the built files at `builtPath` into `outPath`.
+async function linkBuiltFiles(
+  repo: string,
+  builtPath: string,
+  outPath: string
+) {
+  console.log(`Linking built ${repo} into ${outPath}.`);
+  await cleanEmbeddedDir(outPath, repo);
+  if (getOs() === 'windows') {
+    shell.cp('-R', builtPath, p.join(outPath, repo));
+  } else {
+    // Symlinking doesn't play nice with Jasmine's test globbing on Windows.
+    fs.symlink(p.resolve(builtPath), p.join(outPath, repo));
   }
 }
 
-// Ensures that `outPath` exists and its `sass_embedded` subdirectory does not.
-async function cleanEmbeddedDir(outPath: string): Promise<void> {
+// Builds the embedded proto at `protoPath` into a pbjs with TS declaration
+// file.
+function buildEmbeddedProtocol(protoPath: string) {
+  console.log(`Building pbjs and TS declaration file from ${protoPath}.`);
+  try {
+    const protocPath =
+      getOs() === 'windows'
+        ? '%CD%/node_modules/protoc/protoc/bin/protoc.exe'
+        : 'node_modules/protoc/protoc/bin/protoc';
+    const pluginPath =
+      getOs() === 'windows'
+        ? '%CD%/node_modules/.bin/protoc-gen-ts.cmd'
+        : 'node_modules/.bin/protoc-gen-ts';
+    shell.exec(
+      `${protocPath} \
+      --plugin="protoc-gen-ts=${pluginPath}" \
+      --js_out="import_style=commonjs,binary:." \
+      --ts_out="." \
+      ${protoPath}`,
+      {silent: true}
+    );
+  } catch (error) {
+    throw Error(`Failed to write proto to pbjs: ${error.message}.`);
+  }
+}
+
+// Builds the Embedded Dart Sass executable from the source at `repoPath`.
+function buildDartSassEmbedded(repoPath: string) {
+  console.log('Downloading dart-sass-embedded dependencies.');
+  shell.exec('pub upgrade', {
+    cwd: repoPath,
+    silent: true,
+  });
+
+  console.log('Building dart-sass-embedded executable.');
+  shell.exec('pub run grinder protobuf pkg-standalone-dev', {
+    cwd: repoPath,
+    silent: true,
+  });
+}
+
+// Ensures that `outPath` exists and the `subdir` does not.
+async function cleanEmbeddedDir(
```

Comment:
Since this is no longer specific to `sass_embedded`, it should probably have a more generic name.